### PR TITLE
Add parameter freezing `STOLayer` and general performance fix

### PR DESCRIPTION
**Pull request by CheukHinHoJerry**
Updates from this branch (To be updated)

- add freeze parameter `STOLayer`
- type stability fix on some basis
- commented out `FlexArray` interface
- skip bounds check in `SparseSymProd`  inside closure
- prevent calling `generic_matmul!` in `rrule` of `LinearLayer`

cc @DexuanZhou 

@cortner : will let you know once it is ready.

### CODE REVIEW

**Comment by CheukHinHoJerry** [comment]
I think at the end of the PR we should try to generate ALL profview so that we can make sure the time cost go to the right place.

File: src/lux.jl [hunk 1]
```diff
@@ -63,6 +63,7 @@ evaluate!(out, basis::AbstractPoly4MLBasis, X, ps) = evaluate!(out, basis, X)
 function evaluate(l::PolyLuxLayer, X, ps, st)
    out = acquire!(st.pool, _outsym(X), _out_size(l.basis, X), _valtype(l.basis, X))
    evaluate!(out, l.basis, X, ps)
+   release!(X)
```
Comment (cortner) [inline src/lux.jl:66]: flag in LuxLayer on whether or not inputs can be released.

**Review by cortner — CHANGES_REQUESTED** [review]

**Comment by cortner** [comment]
@CheukHinHoJerry  -- the only change since we discussed it is the flag for whether or not to release?

**Comment by CheukHinHoJerry** [comment]
> @CheukHinHoJerry -- the only change since we discussed it is the flag for whether or not to release? I.e. I should just go ahead and merge this?

Yes - I think this is the only thing. And we talked about moving `AtomicorbitalRadius` and wf related stuffs out of this package but I think it should not be done in this PR anyways.

**Pull request merged**